autodocodec/src/Autodocodec/Codec.hs

```diff
@@ -73,53 +75,61 @@ import qualified Data.Aeson.KeyMap as KM
 -- * The @input@ parameter is used for the type that is used during encoding of a value, so it's the @input@ to the codec.
 -- * The @output@ parameter is used for the type that is used during decoding of a value, so it's the @output@ of the codec.
 -- * Both parameters are unused during documentation.
+
+type role Codec _ representational representational
 data Codec context input output where
   -- | Encode '()' to the @null@ value, and decode @null@ as '()'.
-  NullCodec ::
-    ValueCodec () ()
+  NullCodec :: 
+    (Coercible a (), Coercible b ()) => 
+    ValueCodec a b
```

Comment:
I do not understand why this works at all. Could you elaborate?

Reply:
The issue is that `Codec`, as it currently exists, has `nominal` roles for the parameters `input` and `output`. If we look at [Coercible](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Coerce.html#t:Coercible) one can only coerce `T a` to `T b` if `Coercible a b` is true AND the role of the first parameter is representational. And as a consequence, we need the roles of our parameters to be representational if we want to do newtype deriving (as what newtype deriving does is just coercions (which are quite nicely, completely free at runtime).

How do we make the roles of these parameters representational? Well looking at the documentation on [roles](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/roles.html#role-inference) we have this perhaps vague statement:

> all GADT-like parameters are nominal

The docs don't specify what "GADT-like" parameters are, but basically they are, as far as I understand, any GADT  constructor where the RHS isn't completely free in the parameter.

For example:

```
T :: ... -> T a b
```

Doesn't have a GADT style parameter but:

```
T :: ... -> T Int Int
T :: ... -> T a a
T :: C a => ... -> T a b
```

are GADT style.

Any constructor which has a GADT style parameter forces the parameter to be nominal for the whole datatype.

Of course, even if you have:

```
T :: U a b -> T a b
```

And `U`s parameters are nominal, then that will be inherited (this isn't an issue for `Codec`).

So we need to make all the constructors which have values of `input` and `output` which aren't free, to be free.

So for example, with `NullCodec`, instead of `ValueCodec () ()`, we do `(Coercible a (), Coercible b ()) => ValueCodec a b`.

This is really just a generalisation of `NullCodec`, because we could actually write:

```
nullCodec :: (Coercible a (), Coercible b ()) => ValueCodec a b
nullCodec = dimapCodec coerce coerce NullCodec
```

anyway. So doing this refactoring shouldn't break any invariants.

Note that the statement in the [roles docs](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/roles.html#role-inference) that says:

> the default role for classes is nominal

The only exception to this is `Coercible`, which role is representational. 

So by doing this to all the constructors that don't have their parameters free in both parameters, we then make `Codec`'s parameters have representational roles.

Then we can do newtype deriving.

We get some added complexity internally but we give clients more power by giving them the ability to coerce codecs and newtype derive instances.

Including the newtyped instance below.

These would not be possible with the current definition of `Codec`.

Reply:
@clintonmead I'm not convinced yet that this is worth it, but I'm very curious about how this all fits together.
Why can't GHC do this for us?

Reply:
@NorfairKing "Why can't GHC do this for us?"

Because there are good reasons why one would want nominal parameters.

Basically what this transformation does is make the parameters representative, which means for a given data type `T`, makes:

```
Coercible a b => Coercible (T a) (T b)
```

There are some cases where we do not want to do this. Consider the type [Down](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Ord.html#t:Down) and [SortedList](https://hackage.haskell.org/package/sorted-list-0.2.2.0/docs/Data-SortedList.html). 

`Down` exposes it's constructor. In that `coerce :: a -> Down a` is valid.

Should `coerce :: SortedList a -> SortedList (Down a)` be valid?

Certainly not. `SortedList` has an invariant that the elements are sorted based on the `Ord` instance of the type. If you just coerce ` SortedList a -> SortedList (Down a)` the elements are now in the wrong order, violating the invariant of the class.  (Indeed, looking at the source of `SortedList`, I think it's actually buggy and should have a role annotation to prevent this bad coercion). 

There's other circumstance where this transformation just complicate things. For example if you're just making an AST that represents your language, in that you have "Numbers", "Symbols", "Ops" etc. Your list types of expressions are finite and enumerable. You want `Expr Number` or `Expr Symbol` etc, there's no need for the added complexity of `Coercible a Num => Expr a`. 

But I think in the case of `Codec`, there's a few points in favour of this transformation:

1. There's no internal invariants this transformation is violating. This is evidenced by the fact that you could just do `dimap coerce coerce` and achieve the same thing already.
2. Coercing, and in particular, newtype deriving codecs is really useful. Indeed `ToJSON` and `FromJSON` allow for this.